device/blackhole/blackhole_coordinate_manager.cpp

```diff
@@ -1,29 +1,115 @@
 /*
- * SPDX-FileCopyrightText: (c) 2023 Tenstorrent Inc.
+ * SPDX-FileCopyrightText: (c) 2024 Tenstorrent Inc.
  *
  * SPDX-License-Identifier: Apache-2.0
  */
-#include "blackhole_coordinate_manager.h"
-
-std::set<std::size_t> BlackholeCoordinateManager::get_x_coordinates_to_harvest(std::size_t harvesting_mask) {
-    std::set<std::size_t> x_to_harvest;
-    std::size_t logical_x = 0;
-    while (harvesting_mask > 0) {
-        if (harvesting_mask & 1) {
-            x_to_harvest.insert(logical_x);
+#include "umd/device/blackhole_coordinate_manager.h"
+
+using namespace tt::umd;
+
+void BlackholeCoordinateManager::tensix_harvesting(const size_t tensix_harvesting_mask) {
+    CoordinateManager::tensix_harvesting_mask = tensix_harvesting_mask;
+    CoordinateManager::clear_tensix_harvesting_structures();
+
+    size_t num_harvested_x = __builtin_popcount(tensix_harvesting_mask);
+    size_t grid_size_x = CoordinateManager::tensix_grid_size.x;
+    size_t grid_size_y = CoordinateManager::tensix_grid_size.y;
+
+    size_t logical_x = 0;
+    for (size_t x = 0; x < grid_size_x; x++) {
+        if (!(tensix_harvesting_mask & (1 << x))) {
+            for (size_t y = 0; y < grid_size_y; y++) {
+                const tt_xy_pair& tensix_core = CoordinateManager::tensix_cores[x + y * grid_size_x];
+                tensix_logical_to_physical[{logical_x, y}] =
+                    CoreCoord(tensix_core.x, tensix_core.y, CoreType::TENSIX, CoordSystem::PHYSICAL);
+                tensix_physical_to_logical[tensix_core] =
+                    CoreCoord(logical_x, y, CoreType::TENSIX, CoordSystem::LOGICAL);
+            }
+            logical_x++;
+        }
+    }
+
+    for (size_t x = 0; x < grid_size_x - num_harvested_x; x++) {
+        for (size_t y = 0; y < grid_size_y; y++) {
+            const tt_xy_pair& tensix_core = CoordinateManager::tensix_cores[x + y * grid_size_x];
+            tensix_logical_to_virtual[{x, y}] =
+                CoreCoord(tensix_core.x, tensix_core.y, CoreType::TENSIX, CoordSystem::VIRTUAL);
+            tensix_virtual_to_logical[tensix_core] = CoreCoord(x, y, CoreType::TENSIX, CoordSystem::LOGICAL);
+        }
+    }
+
+    BlackholeCoordinateManager::fill_tensix_logical_to_translated();
+}
+
+void BlackholeCoordinateManager::fill_tensix_logical_to_translated() {
+    const size_t num_harvested_x = __builtin_popcount(CoordinateManager::tensix_harvesting_mask);
```

Comment:
This will make the code not cross-compiler friendly. Please write your own implementation (I don't think there's a solution in stdlib) instead, as a private member of coordinate_manager for example.